### PR TITLE
AI Heuristic for Immediate Mode Fallback (TunaNet)

### PR DESCRIPTION
**Pull request by msaudulhassan**
Continuation of: #1837 (too many merge conflicts with develop, so, discontinued #1837, and resuming the work here)

If the cmake variable MIOPEN_ENABLE_AI_HEUR is enabled, Immediate Mode's behavior on a database miss is to use an AI based heurisitic, called TunaNet, to pick the optimal solution. First, the applicability of TunaNet for the given convolution configuration is checked. If TunaNet is applicable, it's fed the parameters of the convolution configuration and it predicts the optimal solver for it (with an accuracy of 90.93%).

**This PR builds on #2031. Reviews on this will be requested once #2031 is complete**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -60,7 +60,7 @@ include(ROCMCreatePackage)
 include(CheckCXXCompilerFlag)
 include(ROCMHeaderWrapper)
 
-
+set(MIOPEN_ENABLE_AI_HEUR On CACHE BOOL "Enable AI based heuristics")
```
Comment (atamazov) [inline CMakeLists.txt]: I highly recommend separating controls for this and for kernel tuning prediction. Specifically, here I recommend using `MIOPEN_ENABLE_AI_IMMEDIATE_MODE_FALLBACK`. See https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2031#discussion_r1149709201

File: include/miopen/config.h.in [hunk 2]
```diff
@@ -49,6 +49,7 @@
 #cmakedefine01 MIOPEN_ENABLE_SQLITE_BACKOFF
 #cmakedefine01 MIOPEN_USE_MLIR
 #cmakedefine01 MIOPEN_USE_COMPOSABLEKERNEL
+#cmakedefine01 MIOPEN_ENABLE_AI_HEUR
```
Comment (atamazov) [inline include/miopen/config.h.in]: Ditto.

File: src/anyramdb.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2022 Advanced Micro Devices, Inc.
```
Comment (atamazov) [inline src/anyramdb.cpp:5]: 2023

**Comment by atamazov** [comment]
@msaudulhassan @JehandadKhan 

> Continuation of: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1837 (too many merge conflicts with develop, so, discontinued https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1837, and resuming the work here)

In #1837:
> Had to create a new PR due to some weird rebasing that occurred when I resolved some merge conflicts in Github's UI.
Link to old PR: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1826

This practice should be stopped. We are losing review comments thus enforcing the reviewers to duplicate their work. Which makes the ~boring~ complex reviewing work ~totally soul-sucking~ even more complex 🤕 

In the future, if you encounter similar problems, please let me know and I will help with resolving merge conflicts.

**Review by atamazov — CHANGES_REQUESTED** [review]
- The reviewed & fixed documentation changes from https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1837/files/5b795cc44064f3977d72318760c1d05756b924ea are missing.

- The PR contains changes from #2031 and therefore can't be reviewed.

**Comment by atamazov** [comment]
@msaudulhassan If this PR depends on #2031, then please indicate this explicitly and mark this as a draft (until #2031 is merged in and this PR is updated from develop). Thanks.

**Comment by msaudulhassan** [comment]
@atamazov I've been having trouble merging `dmantri/convasm_1x1U_KernelTuningNet` into `saud/gfx908_TunaNet`. I tried merging, but now things are in a weird state -- the commits from `saud/gfx908_TunaNet` are now mixed with the commits from `dmantri/convasm_1x1U_KernelTuningNet`, and the files are also mixed. So, I've left this branch and have started a new branch off `dmantri/convasm_1x1U_KernelTuningNet`: https://github.com/ROCmSoftwarePlatform/MIOpen/tree/saud/gfx908_TunaNet_updated . Could you help me merge gfx908_TunaNet_updated into this PR?

**Comment by atamazov** [comment]
@msaudulhassan 
> I tried merging, but now things are in a weird state

If you managed to resolve all merge conflicts, then this draft PR should be fine for now.

>  the commits from saud/gfx908_TunaNet are now mixed with the commits from dmantri/convasm_1x1U_KernelTuningNet

Well, yes, this "mixing" prevents from reviewing code changes made by you. However this is fine as long as you do not need review to begin right away (and you shouldn't need to, because your PR isn't technically ready to be reviewed and merged yet).

- Note: You can prevent mixing of multiple commits when you merge from the other branch. To do so, use the `--squash` option with `git merge`. This should result in a single commit in your branch. That commit shall contain all the commits from the other branch "squashed" into a single one. This technique may be useful in the future.

However, this "mixing" will go away when #2031 is merged into develop and you merge that develop into this branch. When this happens, only your changes will be displayed in this PR and it will be ready for reviewing and merging.

**Pull request closed without merging**